src/main/java/org/testng/JarFileUtils.java

```diff
@@ -57,12 +61,19 @@ private boolean testngXmlExistsInJar(File jarFile, List<String> classes) throws
                 if (matchesXmlPathInJar(je)) {
                     Collection<XmlSuite> parsedSuites = Parser.parse(jf.getInputStream(je), processor);
                     for (XmlSuite suite : parsedSuites) {
-                        // If test names were specified, only run these test names
-                        XmlSuite suiteToAdd = suite;
-                        if (testNames != null) {
-                            suiteToAdd = XmlSuiteUtils.cloneIfContainsTestsWithNamesMatchingAny(suite, testNames);
+                      // If test names were specified, only run these test names
+                      if (testNames != null) {
```

Comment:
Maybe we should add a unit test to `JarFileUtils` which tests this condition ?

Reply:
Here's the source code for it.

[testng-tests.zip](https://github.com/cbeust/testng/files/1529522/testng-tests.zip)

Reply:
Yes, I tried. But finally I figure out we cann't use jar file to test this condition. It would trigger java.io.IOException, which is an existed jdk bug (java 8, 9) 
https://bugs.openjdk.java.net/browse/JDK-815560
https://bugs.openjdk.java.net/browse/JDK-6947916 

You can reproduce easily using my below example.
[testng-tests.zip](https://github.com/cbeust/testng/files/1537349/testng-tests.zip)

Reply:
@eyehwan - I dont understand your comment. What does adding unit tests have anything to do with a bug that talks about `JarURLConnection not handling useCaches` properly ? Can you please elaborate ? 

And if we are not able to write unit tests for this functionality due to the bug that you talk of, I would be curious to know, as to how do you envision this functionality to work for users, who are also using the buggy JDK version ? Because end-users are also going to be providing TestNG with a jar that contains the suite xml in question. 

So if I understand this correctly, then this functionality cannot be used by users as well no ?

Reply:
@krmahadevan , yes, I understand your concern. 

But indeed I have tested this functionality by different runners, command line with jar file, eclipse(IntelliJ) with xml and program. So I do believe it is working well. 

I just have some suspects about way of testing on this case whether it is suitable. You definately have more knowledge than me on this part, if you can help, I would appreciate it. But do you think it is really a stoper on this case?

Reply:
> So I do believe it is working well.

@eyehwan - I was more talking about end-users. One of the ways in which a person makes use of this functionality is via command lines. Lot of folks build uber jars wherein they embed their suite xml in the jar file and then expect TestNG to extract tests from it and work with it. I am merely trying to ensure that, the functionality works the same way across all forms of executing TestNG tests. 

It perhaps would not make a lot of sense for us to tell our users, that this functionality wouldn't work in a particular flavor of execution. At-least that is my opinion on this. 

> I just have some suspects about way of testing on this case whether it is suitable.

Can you please elaborate ? In either case, I think we should be having a test that vets out this functionality via the jars. So if you can give me more details around this, I can try it out over this weekend.

Reply:
>One of the ways in which a person makes use of this functionality is via command lines. Lot of folks build uber jars wherein they embed their suite xml in the jar file and then expect TestNG to extract tests from it and work with it.

Actually this is the way we want to use it in most of time, I guess for other users also  So I spend a lots of time to test this scenario to make it work :)

>Can you please elaborate ? In either case, I think we should be having a test that vets out this functionality via the jars. So if you can give me more details around this, I can try it out over this weekend.

Yes, I will.  If you take look at test example I attached, the mainly differences compared the one you give me is that I add child suite and child of child suite.  The failure would happened at parse xml phase, , I didn't change any test case in JarFileUtilsTest. I attached the stack trace here:
[failure log.txt](https://github.com/cbeust/testng/files/1537663/failure.log.txt)

Reply:
@eyehwan - I still haven't gotten around to this. Kind of swamped by things at work. I was able to reproduce the issue. Here's a workaround that I have been meaning to try. So if you have time, I would request you to please give it a shot. 

Instead of exploding a jar in memory and iterating through each of the entries, feeding them to a parser and then expecting to retrieve the information, can we try filtering out only those entries that a parser can work with ? Currently `org.testng.xml.Parser#getParser` is doing this by looking up all the registered parsers. Once we have a file which our parser can match (assuming that a user didn't inject anything via the SPIs then its always *.xml), we extract out the actual file into the temp folder and then work with those xml files in the temp file system, instead of trying to read them directly from within the jar. 

Yes I know its kind of a round about way of doing this, but this would ensure that we can not only write unit tests for our code, but it would also help people who are for some reason stuck on JDK8 (i guess this got fixed in JDK9) and cant move. Does that make sense ?